### PR TITLE
REST API branch

### PR DESCRIPTION
**Pull request by bryantower**
The goal of this pull request is to add a web server option. You can start the web server and make REST calls to query DiskANN indices. 

There are a few things that should be included before I think it is good enough, but it can safely be merged in its current state. It will not interfere with anything used by the library or the stand alone programs in the current state of things in this PR.

This to do:

- Add more rigorous testing for the web service calls
- Clean up the interface for the REST API
- get a second pair of eyes on the CMake scripts.

I think it can be merged back to main before we get the windows build working. 
- Fix the web server for the windows build. The web server currently does not run properly on Windows.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -17,7 +17,11 @@
 # Some variables like MSVC are defined only after project(), so put that first.
 project(diskann)
 
-cmake_minimum_required(VERSION 3.15)
+if(MSVC)
```
Comment (daxpryce) [inline CMakeLists.txt:20]: I say we just make this a minimum of 3.15 across the board. Why use such an old version of cmake?

**Comment by daxpryce** [comment]
Wasn't there a version of this branch where we were using the boost program argument parser instead of just positional arguments?  I could swear we had that at one point. I don't think it has to come in with this PR, I was just surprised not to see it :)

File: CMakeLists.txt (hunk 1 above)
Comment (bryantower) [reply CMakeLists.txt]: I agree. I will just simplify this to make it 3.15 everywhere.

**Comment by bryantower** [comment]
> Wasn't there a version of this branch where we were using the boost program argument parser instead of just positional arguments? I could swear we had that at one point. I don't think it has to come in with this PR, I was just surprised not to see it :)

There may have been, but I don't think I saw it. I can add that in, but I agree it doesn't need to be in this PR.

File: include/restapi/server.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+#include <restapi/common.h>
+#include <cpprest/http_listener.h>
+
+namespace diskann {
+  class Server {
+   public:
+    // Server(web::uri& url, std::unique_ptr<diskann::BaseSearch>& searcher,
+    //        const std::string& typestring);
+    Server(web::uri&                                          url,
```
Comment (harsha-simhadri) [inline include/restapi/server.h:14]: remove commented code

**Review by harsha-simhadri — CHANGES_REQUESTED** [review]

File: src/restapi/server.cpp [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+#include <iomanip>
+#include <string>
+#include <cstdlib>
+#include <codecvt>
+#include <limits>
+
+#include <restapi/server.h>
+
+namespace diskann {
+  const unsigned int DEFAULT_L = 100;
```
[the remark this answers is not in the document]
Comment (bryantower) [reply src/restapi/server.cpp:15]: Good idea. I will move this to the common.h

File: tests/restapi/client.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            std::wcout << e.what() << std::endl;
+          }
+        })
+        .wait();
+  }
+}
+
+int main(int argc, char* argv[]) {
+  if (argc != 6) {
+    std::cout << "Usage: client ip_addr:port data_type<float/int8/uint8> query_file "
```
[the remark this answers is not in the document]
Comment (bryantower) [reply tests/restapi/client.cpp]: added options to change distance metrics

File: tests/restapi/inmem_server.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+              << std::endl;
+    exit(1);
+  }
+
+  std::string address(argv[1]);
+  const char* data_file = argv[3];
+  const char* index_file = argv[4];
+  const unsigned num_threads = atoi(argv[5]);
+  const unsigned l_search = atoi(argv[6]);
+  const char* tags_file = argc == 8 ? argv[7] : nullptr;
```
[the remark this answers is not in the document]
Comment (bryantower) [reply tests/restapi/inmem_server.cpp]: added options to change distance metrics

File: src/restapi/server.cpp [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+                         int64_t& queryId, T*& queryVector,
+                         unsigned int& dimensions, unsigned& Ls) {
+    std::cout << body << std::endl;
+    web::json::value val = web::json::value::parse(body);
+    web::json::array queryArr = val.at(VECTOR_KEY).as_array();
+    queryId = val.has_field(QUERY_ID_KEY)
+                  ? val.at(QUERY_ID_KEY).as_number().to_int64()
+                  : -1;
+    Ls = val.has_field(L_KEY) ? val.at(L_KEY).as_number().to_uint32()
+                              : DEFAULT_L;
```
[the remark this answers is not in the document]
Comment (bryantower) [reply src/restapi/server.cpp:178]: enforcing this now.

File: tests/python/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+numpy
```
[the remark this answers is not in the document]
Comment (bryantower) [reply tests/python/requirements.txt:1]: I have renamed the folder to tests/python/restapi, that should leave us plenty of room to expand to other python tests if needed.

File: tests/python/README.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
+# Integration Tests
+The following tests use Python to prepare, run, verify, and tear down the rest api services.
+
+We do make use of the built-in `unittest` library, but that's only to take advantage of test reporting purposes.
+
```
[the remark this answers is not in the document]
Comment (bryantower) [reply tests/python/README.md:8]: I have enabled the REST_API flag on Linux, there is an issue with character sets on windows. I really think that we should acknowledge that the Windows REST API is not working and proceed with the merge. After the merge we should make an issue for the windows server and figure out the priority of a windows server.

File: src/restapi/server.cpp [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+          try {
+            T*           queryVector = nullptr;
+            unsigned int dimensions = 0;
+            unsigned int Ls;
+            parseJson(body, K, queryId, queryVector, dimensions, Ls);
+
+            auto startTime = std::chrono::high_resolution_clock::now();
+            std::vector<diskann::SearchResult> results;
+
+            for (auto& searcher : _multi_searcher)
```
[the remark this answers is not in the document]
Comment (bryantower) [reply src/restapi/server.cpp:115]: I don't know if making this multi-threaded will help too much. Each search is already multi-threaded. Will the two layers interact with each other?

**Review by harsha-simhadri — APPROVED** [review]

**Pull request merged**